### PR TITLE
[Data] [1/n] Async `iter_batches`: Add Threadpool util

### PR DESCRIPTION
**Pull request by amogkam**
Part 1 of async `iter_batches` support. Converts the existing `_make_async_gen` util to be able to use a threadpool instead of a single thread.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/_internal/block_batching/util.py
Comment (ericl) [inline python/ray/data/_internal/block_batching/util.py:71]: Question about this block=True flag here. Since this is run on the main thread, could it block the entire main thread unnecessarily?

For example, suppose we chained three _make_async_gen calls, and the last generator blocked on getting an output. Then it seems to me that the intermediate generators couldn't yield items to the worker threads to consume just because the last one was blocked.

**Review by ericl — APPROVED** [review]

**Comment by ollie-iterators** [comment]
It looks like there is an issue on line 169 of the file "python/ray/data/_internal/block_batching/block_batching.py". The text "prefetch_buffer_size" should be removed as it is not a part of the _make_async_gen function anymore.

File: python/ray/data/_internal/block_batching/util.py
Comment (jianoaix) [inline python/ray/data/_internal/block_batching/util.py:20]: The last sentence seems missing some part?

File: python/ray/data/_internal/block_batching/util.py
Comment (jianoaix) [inline python/ray/data/_internal/block_batching/util.py:29]: This ("same elements") seems no longer true since there is a `fn` to transform the inputs.

File: python/ray/data/_internal/block_batching/util.py
Comment (amogkam) [reply python/ray/data/_internal/block_batching/util.py]: good catch, updated

File: python/ray/data/_internal/block_batching/util.py
Comment (amogkam) [reply python/ray/data/_internal/block_batching/util.py]: good catch, updated

File: python/ray/data/_internal/block_batching/util.py
Comment (clarkzinzow) [inline python/ray/data/_internal/block_batching/util.py:86]: This blocks the calling thread until all work is done, but the calling thread is the one doing work (i.e. it is the one getting from the queue and calling `.task_done()`), so I don't think that this `output_queue.join()` call makes sense here?

File: python/ray/data/_internal/block_batching/util.py
Comment (clarkzinzow) [inline python/ray/data/_internal/block_batching/util.py:81]: Nit: Any reason that we couldn't join all of the threads at once upon receiving `num_workers` sentinel objects, indicating that all worker threads are done? Then the sentinel object could be something a lot simpler like `object()`.

File: python/ray/data/_internal/block_batching/block_batching.py
Comment (clarkzinzow) [inline python/ray/data/_internal/block_batching/block_batching.py:168]: Hmm so this wrapping `_iterator_fn` is necessary because in `_make_async_gen` we want to synchronize the `next(blocks)` but _not_ the `next(batch_iter)`, since the latter is the expensive bit (batching, formatting) that we want to parallelize? Might want to leave that as a comment to make that clear.

File: python/ray/data/_internal/block_batching/util.py
Comment (clarkzinzow) [inline python/ray/data/_internal/block_batching/util.py:21]: To make it extra clear where the concurrent/parallel computation will happen.
```suggestion
    thread-safe fashion, and apply the provided computation concurrently
    (and in parallel, for the bits of fn that release the GIL).
```

**Review by clarkzinzow — COMMENTED** [review]
Mostly nits/questions; my biggest questions are:
1. Whether making the threads daemonic will lead to improperly released resources, which might lead to some bad bugs.
2. Whether `output_queue.join()` should be called from the thread that calls `.task_done()`, which I think might be nonsensical.

**Comment by ericl** [comment]
@clarkzinzow regarding daemon threads, the main thing is not leaking drivers / worker processes when things aren't properly shut down. For non-daemon threads, those hold the entire process open until you force exit.

File: python/ray/data/_internal/block_batching/util.py
Comment (amogkam) [reply python/ray/data/_internal/block_batching/util.py:81]: Yeah this could also work, but opted for immediately terminating each thread as soon as that thread has finished.

File: python/ray/data/_internal/block_batching/util.py
Comment (amogkam) [reply python/ray/data/_internal/block_batching/util.py:86]: Hmm, not sure I'm following. If we hit this `if` statement, then all work has been already finished, since all threads have completed and all items have been yielded already.

File: python/ray/data/_internal/block_batching/util.py
Comment (amogkam) [reply python/ray/data/_internal/block_batching/util.py:64]: In this case, the event is only set if one of the child threads raises an exception. But what happens if the main thread is interrupted? In that case, the event will not be set and the child threads will still be alive right?

File: python/ray/data/_internal/block_batching/block_batching.py
Comment (amogkam) [reply python/ray/data/_internal/block_batching/block_batching.py:168]: Yep that's right!

Comment (amogkam) [reply python/ray/data/_internal/block_batching/block_batching.py:168]: I actually think the current approach of passing in a separate `fn` to execute is more intuitive.

**Comment by clarkzinzow** [comment]
@ericl Ah, so we're making the worker threads daemonic to prevent spurious blocking within `fn` (which would circumvent our shutdown event checking) from keeping the process from exiting, and we're only terminating these threads asynchronously when the process is exiting so we don't care about properly releasing resources. Got it, sounds good!

File: python/ray/data/_internal/block_batching/util.py
Comment (clarkzinzow) [reply python/ray/data/_internal/block_batching/util.py:86]: Yeah, I'm saying that this `output_queue.join()` is nonsensical, since it should never block, and if it ever did (e.g. due to a bug), it would block forever.

All that `Queue.join()` does is block until all work is done, i.e. for every item put in the queue, an `output_queue.task_done()` call has been made; this is why this is typically called in the producer thread (the thread calling `output_queue.put()`), since it's waiting for the consumer(s) to be done doing work. But here we're calling it in the consumer thread (the thread calling `output_queue.get()` and `output_queue.task_done()`), and if `output_queue.join()` ever blocks, it would block forever, since no other threads can call `output_queue.task_done()` to unblock it.

I'm saying that I think we should remove the `output_queue.join()` call here. Does that make sense?

Comment (amogkam) [reply python/ray/data/_internal/block_batching/util.py:90]: Ah got it

File: python/ray/data/_internal/block_batching/block_batching.py
Comment (clarkzinzow) [reply python/ray/data/_internal/block_batching/block_batching.py:168]: Agreed, it's definitely less ambiguous than trying to determine in which thread the implicit `next()` is being called, e.g. when iterating over an iterator via a for loop.

File: python/ray/data/_internal/block_batching/util.py
Comment (amogkam) [reply python/ray/data/_internal/block_batching/util.py]: removed

File: python/ray/data/_internal/block_batching/block_batching.py
Comment (amogkam) [reply python/ray/data/_internal/block_batching/block_batching.py:168]: added comment

File: python/ray/data/_internal/block_batching/util.py
Comment (amogkam) [reply python/ray/data/_internal/block_batching/util.py]: updated

**Review by clarkzinzow — APPROVED** [review]

**Pull request merged**
